### PR TITLE
cpu: lpc11u34: Added new CPU defintiion for NXP LPC11U34

### PR DESCRIPTION
**Pull request by ks156**
CPU definition for [NXP LPC11U34](http://www.nxp.com/documents/data_sheet/LPC11U3X.pdf), a Cortex-M0 CPU.
The peripherals already added are : 

- UART
- timer 
- GPIO. 

All the peripherals have been tested and are working.

This CPU definition is the first step for adding support for the [WeIO](http://www.we-io.net) board.

### CODE REVIEW

**Comment by OlegHahm** [comment]
Welcome to RIOT and thanks for your contribution. Seems to be an interesting board. I guess you're planning to run RIOT on the M0 only (and not on the bigger ARM), right?

**Comment by ks156** [comment]
Hello @OlegHahm,
For now, WeIO is in its first version with a Cortex M0 MCU. We want to use RIOT to replace the existing firmware. For a second version of the board, it's quite possible that we'll switch to a STM MCU, M0 or M3.

**Comment by ks156** [comment]
In parallel with the CPU definition, I'm writing the board (weio) support. 
Should I add the board also in this pull request, or is it preferable to create another PR ?

**Comment by OlegHahm** [comment]
I think a separate PR for the board would be cleaner and better to review.

**Comment by drasko** [comment]
@OlegHahm this is correct, except it is not bigger ARM, but MIPS :). These MIPS Linux CPU-s with WiFi integrated are quite usual in router world and OpenWrt Linux distro is running on the CPU.

Thanks to the huge quantities of these devices produced in China, there is a lot of sub-$10 FCC/CE certified WiSoC modules (WiFi integrated Linux capable CPU + DDRAM + flash). For example, WeIO is based on this one: http://www.8devices.com/carambola-2, but today you can find much cheaper and more powerful ones.

As a consequence, this today represents probably the cheaper way to make a Linux (industrial) computer, if we exclude shady CHIP (https://www.kickstarter.com/projects/1598272670/chip-the-worlds-first-9-computer) or RPi deals, or subventioned Beagle Board...

However - these modules are intended to be used for building routers - which means that they have very limited number of GPIOs and not at all ADC or PWM units. This is why we added small Cortex-M0 - it serves as an "expander", to add missing RT peripherals.

We have developed simple TLV protocol for inter-processor communication, but we find that RIOT can actually be an excellent HAL for what we need - and that is more tight RPC protocol based on CBOR. If we achieve this, it can be very interesting and generic way to build low-cost but powerful WiFi connected Linux modules, as we see that more and more boards that appearing are adopting WeIO architecture (https://tessel.io/, http://wrtnode.com/w/?p=671).

**Comment by OlegHahm** [comment]
Ah, it says "AR9331" and I added an `M` after the `R`. ;)

**Comment by thomaseichinger** [comment]
@ks156 Welcome and great to see this port! Even though easier to review (less files) I propose to add the board to this PR, else the CI system won't pick up this code.

**Comment by ks156** [comment]
Hello @thomaseichinger, 
I pushed the board definition.

**Comment by ks156** [comment]
@thomaseichinger : I see that both tests failed.
I see issues with headers that I can fix, but when I see the Travis CI logs, I don't understand why cortex_m0 has failed. I don't see any output that can give me an advice about the problem.

Is there a way to run these tests locally, or without pushing code into this PR ?

**Comment by OlegHahm** [comment]
@ks156, only static tests are run by default. Build tests, unittests, and the like are only performed if a maintainer sets the appropriate label. However, the static tests already show one tiny glitch:
> file has an unknown license header: 'boards/weio/system.c'

I think you should be able to run the tests locally with the script in `dist/tools/travis-scripts/build_and_test.sh`.

**Comment by ks156** [comment]
Ok, thank you for this answer. I'll fix the header tomorrow morning, and try to use the script to perform the test.

**Comment by thomaseichinger** [comment]
@ks156 Regarding `system.c`, we try to avoid the CMSIS `SystemInit` functions as from the whole file there are only a couple of lines actually used. This way we don't pull in a lot of unused code and it also solves the unknown license problem.

**Comment by ks156** [comment]
@thomaseichinger I rewrite ````system.c```` with only the board specific configuration. The file is now clean.

However, tests still fail due to Doxygen warning in ````LPC11Uxx.h````. When I run Doxygen locally (```make doc```), I don't see any warning for this file..

**Comment by OlegHahm** [comment]
```
make doc 2>&1 | grep LPC
Preprocessing /home/oleg/git/RIOT/cpu/lpc11u34/include/LPC11Uxx.h...
Parsing file /home/oleg/git/RIOT/cpu/lpc11u34/include/LPC11Uxx.h...
Generating code for file LPC11Uxx.h...
Generating docs for file l/home/oleg/git/RIOT/cpu/lpc11u34/include/LPC11Uxx.h:611: warning: Member LPC_I2C_BASE (macro definition) of group Device_Peripheral_Registers is not documented.
/home/oleg/git/RIOT/cpu/lpc11u34/include/LPC11Uxx.h:612: warning: Member LPC_WWDT_BASE (macro definition) of group Device_Peripheral_Registers is not documented.
/home/oleg/git/RIOT/cpu/lpc11u34/include/LPC11Uxx.h:613: warning: Member LPC_USART_BASE (macro definition) of group Device_Peripheral_Registers is not documented.
/home/oleg/git/RIOT/cpu/lpc11u34/include/LPC11Uxx.h:614: warning: Member LPC_CT16B0_BASE (macro definition) of group Device_Peripheral_Registers is not documented.
```
and several more warnings of this kind. Since these files are not really part of the OS, but only documentation of the hardware, it would be also okay to add these files to the exclude list in `doc/doxygen/riot.doxyfile`.

**Comment by ks156** [comment]
I've fixed the doxygen documentation in LPC11Uxx.h. When I run doxygen (1.8.10), I don't have warnings anymore. But tests still fail. So I'll add this file in the exclude list.

**Comment by ks156** [comment]
I tried to squash the commit as asked in the test results, but it looks like I did something wrong.
I've read this doc : https://github.com/RIOT-OS/RIOT/wiki/Git-cheatsheet to understand how to do, but still...
Someone can help me ?

File: boards/weio/Makefile.include [hunk 1]
```diff
@@ -0,0 +1,21 @@
+# define the cpu used by the mbed_lpx1768 board
```
Comment (thomaseichinger) [inline boards/weio/Makefile.include]: s/mbed_lpx1768/weio/

File: boards/weio/Makefile.include [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+export CPU = lpc11u34
+
+export FLASHER = $(RIOTBOARD)/$(BOARD)/dist/flash.sh
```
Comment (thomaseichinger) [inline boards/weio/Makefile.include]: This file is not included in this PR.

File: boards/weio/board.c [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+/**
+ * @brief Initialize the boards on-board LEDs (LED1 to LED4)
+ *
+ * The LED initialization is hard-coded in this function. As the LEDs are soldered
+ * onto the board they are fixed to their CPU pins.
+ *
+ * The LEDs are connected to the following pins:
+ * - LED1: P1.18
+ * - LED2: P1.19
+ * - LED3: P1.20
```
Comment (thomaseichinger) [inline boards/weio/board.c]: Are these correct? In `board.h` `BIT13, BIT14, BIT15` are used to define `LED?_PIN`.

File: boards/weio/include/periph_conf.h [hunk 4]
```diff
@@ -0,0 +1,213 @@
...
+ * Public License v2.1. See the file LICENSE in the top level directory for more
+ * details.
+ */
+
+/**
+ * @ingroup     boards_weio
+ * @{
+ *
+ * @file
+ * @brief       Peripheral MCU configuration for the mbed LPC1768 board
```
Comment (thomaseichinger) [inline boards/weio/include/periph_conf.h]: Wrong board.

File: boards/weio/include/periph_conf.h [hunk 5]
```diff
@@ -0,0 +1,213 @@
...
+ *
+ * @author      Hauke Petersen <hauke.petersen@fu-berlin.de>
```
Comment (thomaseichinger) [inline boards/weio/include/periph_conf.h]: Don't forget to add yourself as author.

File: boards/weio/include/periph_conf.h [hunk 6]
```diff
@@ -0,0 +1,213 @@
...
+#define UART_0_TX_PINMODE   (LPC_IOCON->PIO0_19)
+#define UART_0_RX_PINMODE   (LPC_IOCON->PIO0_18)
+#define UART_0_TX_PIN       (1)
+#define UART_0_RX_PIN       (2)
+#define UART_0_AF           (1)
+/** @} */
+
+/**
+ *  * @name GPIO configuration
+ *   * @{
```
Comment (thomaseichinger) [inline boards/weio/include/periph_conf.h]: Please remove extra `  *`.

File: boards/weio/system.c [hunk 7]
```diff
@@ -0,0 +1,119 @@
...
+/* Define clocks */
+#define __XTAL            (12000000UL)    /* Oscillator frequency */
+#define __SYS_OSC_CLK     (    __XTAL)    /* Main oscillator frequency */
+
+#define __SYS_PLLCLKIN    (__SYS_OSC_CLK)
+#define  __SYS_PLLCLKOUT  (__SYS_PLLCLKIN * ((SYSPLLCTRL_Val & 0x01F) + 1))
+#define __MAIN_CLOCK      (__SYS_PLLCLKOUT)
+#define __SYSTEM_CLOCK    (__MAIN_CLOCK / SYSAHBCLKDIV_Val)
+
+uint32_t SystemCoreClock = __SYSTEM_CLOCK; /*!< System Clock Frequency (Core Clock)*/
```
Comment (thomaseichinger) [inline boards/weio/system.c]: This whole cascade of defines doesn't seem to be used.

File: boards/weio/board.c (hunk 3 above)
Comment (ks156) [inline boards/weio/board.c]: Indeed, not correct. I'll correct it.

File: boards/weio/system.c [hunk 8]
```diff
@@ -0,0 +1,119 @@
...
+/**
+ * Initialize the system
+ *
+ * @param  none
+ * @return none
+ *
+ * @brief  Setup the microcontroller system.
+ *         Initialize the System.
+ */
+void SystemInit (void) {
```
Comment (thomaseichinger) [inline boards/weio/system.c]: I didn't look into it in detail yet, is this clock setup highly specific to the WeIO board? If not, I would propose to move it to `cpu.c` to `clk_init()` as it is done for most other boards/CPUs.

File: boards/weio/system.c [hunk 9]
```diff
@@ -0,0 +1,119 @@
...
+    /* Toggle Update Register */
+    LPC_SYSCON->USBCLKUEN = 0x00;
+    LPC_SYSCON->USBCLKUEN = 0x01;
+    /* Set USB clock divider */
+    LPC_SYSCON->USBCLKDIV = USBCLKDIV_Val;
+
+    /* System clock to the IOCON needs to be enabled or
+       most of the I/O related peripherals won't work. */
+    LPC_SYSCON->SYSAHBCLKCTRL |= (1<<16);
+
```
Comment (thomaseichinger) [inline boards/weio/system.c]: Please remove extra newline.

File: cpu/lpc11u34/periph/gpio.c [hunk 10]
```diff
@@ -0,0 +1,503 @@
...
+#include "periph/gpio.h"
+
+typedef struct {
+    gpio_cb_t cb;
+    void *arg;
+} gpio_state_t;
+
+/* Static IOCON registers definition */
+volatile uint32_t * const LPC_PIN_REGISTERS[] = {
+    // PORT 0 (PIO0_0 -> PIO0_23)
```
Comment (thomaseichinger) [inline cpu/lpc11u34/periph/gpio.c]: Please use `/*  */`

**Comment by OlegHahm** [comment]
> I tried to squash the commit as asked in the test results, but it looks like I did something wrong.

I see how the test output might be misleading, but according to point 9 in the [development procedures](https://github.com/RIOT-OS/RIOT/wiki/Development-procedures) you should wait before squashing until the reviewer tells you so.

> I've read this doc : https://github.com/RIOT-OS/RIOT/wiki/Git-cheatsheet to understand how to do, but still...
Someone can help me ?

Can you elaborate on which problems you're having?

File: cpu/lpc11u34/periph/gpio.c [hunk 11]
```diff
@@ -0,0 +1,503 @@
...
+#include "thread.h"
+#include "periph/gpio.h"
+
+typedef struct {
+    gpio_cb_t cb;
+    void *arg;
+} gpio_state_t;
+
+/* Static IOCON registers definition */
+volatile uint32_t * const LPC_PIN_REGISTERS[] = {
```
Comment (thomaseichinger) [inline cpu/lpc11u34/periph/gpio.c]: All capitals should only be used for macros.

File: cpu/lpc11u34/periph/gpio.c [hunk 12]
```diff
@@ -0,0 +1,503 @@
...
+    port = gpio_port_map[dev];
+    pin = gpio_pin_map[dev];
+
+    /* Put pin in the primary function */
+    *LPC_PIN_REGISTERS[pin + (port * 24)] = *LPC_PIN_REGISTERS[pin + (port * 24)] & ~(0x07);
+
+    /* Disable resistors */
+    *LPC_PIN_REGISTERS[pin + (port * 24)] &= ~(3 << 3);
+    /* Set resistors */
+    if (pullup == GPIO_PULLUP)
```
Comment (thomaseichinger) [inline cpu/lpc11u34/periph/gpio.c]: Please use `{}` here.

File: cpu/lpc11u34/periph/gpio.c [hunk 13]
```diff
@@ -0,0 +1,503 @@
...
+        *LPC_PIN_REGISTERS[pin + (port * 24)] |= (2 << 3);
+    else if (pullup == GPIO_PULLDOWN)
```
Comment (thomaseichinger) [inline cpu/lpc11u34/periph/gpio.c]: and here

File: cpu/lpc11u34/periph/gpio.c [hunk 14]
```diff
@@ -0,0 +1,503 @@
...
+        *LPC_PIN_REGISTERS[pin + (port * 24)] |= (1 << 3);
+
+    /* Set direction */
+    if (dir == GPIO_DIR_OUT) {
+        LPC_GPIO->DIR[port] |= (1 << pin);           /* set pin to output mode */
+    }
+    else {
+        LPC_GPIO->DIR[port] &= ~(1 << pin);           /* set pin to output mode */
+
```
Comment (thomaseichinger) [inline cpu/lpc11u34/periph/gpio.c]: Remove extra line.

File: boards/weio/system.c (hunk 8 above)
Comment (ks156) [inline boards/weio/system.c]: This setup is indeed not highly specific to WeIO, but configure the MCU to use an external crystal at 12MHz. This config will not be usable for boards that use internal RC for example.

File: cpu/lpc11u34/periph/gpio.c [hunk 15]
```diff
@@ -0,0 +1,503 @@
...
+    uint8_t pin;
+    uint8_t port;
+
+    LPC_SYSCON->SYSAHBCLKCTRL |= (1 << 19);
+
+    if (dev >= GPIO_NUMOF) {
+        return -1;
+    }
+
+    if (gpio_int_id > 7)
```
Comment (thomaseichinger) [inline cpu/lpc11u34/periph/gpio.c]: Please add `{}`s.

File: cpu/lpc11u34/periph/gpio.c [hunk 16]
```diff
@@ -0,0 +1,503 @@
...
+    LPC_SYSCON->PINTSEL[gpio_int_id] = (port * 24) + pin;
+
+    /* set callback */
+    gpio_config[gpio_int_id].cb = cb;
+    gpio_config[gpio_int_id].arg = arg;
+
+    /* configure the event that triggers an interrupt */
+    switch (flank) {
+        case GPIO_RISING:
+            LPC_GPIO_PIN_INT->ISEL &= ~(1 << gpio_int_id); // Set PMODE=edge sensitive
```
Comment (thomaseichinger) [inline cpu/lpc11u34/periph/gpio.c]: Here and following, please change to use `/*  */`.

File: cpu/lpc11u34/periph/gpio.c [hunk 17]
```diff
@@ -0,0 +1,503 @@
...
+    }
+
+    int_id = flex_int_mapping[dev] - 1;
+
+    NVIC_DisableIRQ(int_id);
+    LPC_GPIO_PIN_INT->CIENR = (1 << int_id);   // Disable rising edge or level interrupt
+    LPC_GPIO_PIN_INT->CIENF = (1 << int_id);    // Disable falling edge interrupt
+    LPC_GPIO_PIN_INT->RISE = (1 << int_id);    // Clear rising edge (sort of) flag
+    LPC_GPIO_PIN_INT->FALL = (1 << int_id);    // Clear falling edge (sort of) flag
+}
```
Comment (thomaseichinger) [inline cpu/lpc11u34/periph/gpio.c]: Calling `gipo_irq_disable` currently omits interrupt edge configuration. Wouldn't it be sufficient to call `NVIC_DisableIRQ`? This way a consecutive call to `gpio_irq_enable` could resume with the init configuration.

Comment (thomaseichinger) [inline cpu/lpc11u34/periph/gpio.c]: Oh, and please use `/*  */`.

File: cpu/lpc11u34/periph/gpio.c [hunk 18]
```diff
@@ -0,0 +1,503 @@
...
+    if (value) {
+        gpio_set(dev);
+    } else {
+        gpio_clear(dev);
+    }
+}
+
+static inline void isr_common(uint8_t int_id) {
+
+    LPC_GPIO_PIN_INT->IST |= (1 << int_id); // Clear pending interrupt
```
Comment (thomaseichinger) [inline cpu/lpc11u34/periph/gpio.c]: `//` -> `/*  */`

File: cpu/lpc11u34/periph/timer.c [hunk 19]
```diff
@@ -0,0 +1,191 @@
...
+    if (TIMER_0_DEV->IR & MR3_FLAG) {
+        TIMER_0_DEV->IR |= (MR3_FLAG);
+        TIMER_0_DEV->MCR &= ~(1 << 9);
+        config[TIMER_0].cb(3);
+    }
+    if (sched_context_switch_request) {
+        thread_yield();
+    }
+}
+#endif
```
Comment (thomaseichinger) [inline cpu/lpc11u34/periph/timer.c]: This is already guarded by `#ifdef TIMER_0_EN`. (See line below)

**Comment by ks156** [comment]
>I tried to squash the commit as asked in the test results, but it looks like I did something wrong.
I see how the test output might be misleading, but according to point 9 in the development procedures you should wait before squashing until the reviewer tells you so.

Ok, sorry. I misinterpreted the test output. Next time I'll wait before squashing.

>Can you elaborate on which problems you're having?

I don't know exactly. The output looks strange, and I don't know if things happened correctly / as expected, or not. I'm a noob with  the *rebase* function.

File: cpu/lpc11u34/vectors.c [hunk 20]
```diff
@@ -0,0 +1,120 @@
...
+WEAK_DEFAULT void isr_bod(void);
+WEAK_DEFAULT void isr_flash(void);
+WEAK_DEFAULT void isr_usb_wakeup(void);
+
+/* interrupt vector table */
+__attribute__ ((section(".vectors")))
+const void *interrupt_vector[] = {
+    /* Exception stack pointer */
+    (void*) (&_estack),             /* pointer to the top of the stack */
+    /* Cortex-M3 handlers */
```
Comment (thomaseichinger) [inline cpu/lpc11u34/vectors.c]: s/M3/M0/

File: cpu/lpc11u34/vectors.c [hunk 21]
```diff
@@ -0,0 +1,120 @@
...
+    (void*) isr_pendsv,             /* pendSV interrupt, in RIOT the actual
+                                     * context switching is happening here */
+    (void*) isr_systick,            /* SysTick interrupt, not used in RIOT */
+    /* LPC specific peripheral handlers */
+    (void*) isr_pinint0,            /* Pin ISR0 */
+    (void*) isr_pinint1,            /* Pin ISR1 */
+    (void*) isr_pinint2,            /* Pin ISR2 */
+    (void*) isr_pinint3,            /* Pin ISR3 */
+    (void*) isr_pinint4,            /* Pin ISR4 */
+    (void*) isr_pinint5,	    /* Pin ISR5 */
```
Comment (thomaseichinger) [inline cpu/lpc11u34/vectors.c]: Indention.

File: cpu/lpc11u34/periph/gpio.c (hunk 17 above)
Comment (ks156) [inline cpu/lpc11u34/periph/gpio.c]: Indeed, NVIC_DisableIRQ is sufficient here.

**Comment by thomaseichinger** [comment]
@ks156 Great job. Some minor comments, mostly on style and documentation. 
Btw, how do you flash this board?

File: boards/weio/system.c (hunk 8 above)
Comment (thomaseichinger) [inline boards/weio/system.c]: Ok, I would still vote to take this into cpu for now. We can then think about a way to separate this once there is a second board using this MCU.

**Comment by ks156** [comment]
>@ks156 Great job. Some minor comments, mostly on style and documentation. 

Thank you. I'll fix all your remarks quickly.

>Btw, how do you flash this board?

The MCU is connected to a MIPS (AR9331) module, via USB. LPC11U34 has a USB  bootloader, and we flash it this way.
The firmware must be sent to the AR9331 (OpenWRT), and then a script allows to active the bootloader and flash the firmware.

I have to write the script. The binary file will probably be sent via *scp*, and the commands to flash the MCU via *ssh*.

File: cpu/lpc11u34/periph/timer.c (hunk 19 above)
Comment (ks156) [inline cpu/lpc11u34/periph/timer.c]: Removed the protection around the ISR handler

File: boards/weio/system.c (hunk 8 above)
Comment (ks156) [inline boards/weio/system.c]: Ok, I'll do that way. ```system.c``` is not longer required then. The file can be removed ?

Comment (thomaseichinger) [inline boards/weio/system.c]: > The file can be removed ?
Yes, `system.c` can be removed then.

Comment (ks156) [inline boards/weio/system.c]: ok, moved into ```cpu.c```, clk_init()

**Comment by ks156** [comment]
@thomaseichinger I've finished to fix all your remarks. Thank you for this review.

**Comment by thomaseichinger** [comment]
@ks156 Some tests/examples are failing because of insufficient RAM/ROM. You will need to black list them in the according Makefiles.

**Comment by ks156** [comment]
>@ks156 Some tests/examples are failing because of insufficient RAM/ROM. You will need to black list them in the according Makefiles.

Ok, I'll check how to do that

I suppose that the blacklist are performed by adding the board here : https://github.com/RIOT-OS/RIOT/blob/master/tests/slip/Makefile#L4 for example ?

**Comment by thomaseichinger** [comment]
> I suppose that the blacklist are performed by adding the board here : https://github.com/RIOT-OS/RIOT/blob/master/tests/slip/Makefile#L4 for example ?

Exactly.

**Comment by ks156** [comment]
Excepted the static tests, the other tests seems OK \o/

File: cpu/lpc11u34/cpu.c [hunk 22]
```diff
@@ -0,0 +1,115 @@
...
+
+#define SYSOSCCTRL_Val    0x00000000 // Reset: 0x000
+#define SYSPLLCTRL_Val    0x00000023 // Reset: 0x000
+#define SYSPLLCLKSEL_Val  0x00000001 // Reset: 0x000
+#define MAINCLKSEL_Val    0x00000003 // Reset: 0x000
+#define SYSAHBCLKDIV_Val  0x00000001 // Reset: 0x001
+#define USBPLLCTRL_Val    0x00000023 // Reset: 0x000
+#define USBPLLCLKSEL_Val  0x00000001 // Reset: 0x000
+#define USBCLKSEL_Val     0x00000000 // Reset: 0x000
+#define USBCLKDIV_Val     0x00000001 // Reset: 0x001
```
Comment (thomaseichinger) [inline cpu/lpc11u34/cpu.c]: `//` -> `/*  */`

File: cpu/lpc11u34/cpu.c [hunk 23]
```diff
@@ -0,0 +1,115 @@
...
+    /* Toggle Update Register */
+    LPC_SYSCON->USBCLKUEN = 0x00;
+    LPC_SYSCON->USBCLKUEN = 0x01;
+    /* Set USB clock divider */
+    LPC_SYSCON->USBCLKDIV = USBCLKDIV_Val;
+
+    /* System clock to the IOCON needs to be enabled or
+       most of the I/O related peripherals won't work. */
+    LPC_SYSCON->SYSAHBCLKCTRL |= (1<<16);
+
```
Comment (thomaseichinger) [inline cpu/lpc11u34/cpu.c]: Please remove extra newline.

File: cpu/lpc11u34/periph/gpio.c [hunk 24]
```diff
@@ -0,0 +1,501 @@
...
+    LPC_GPIO->CLR[port] = (1 << pin);
+}
+
+void gpio_toggle(gpio_t dev)
+{
+    if (gpio_read(dev)) {
+        gpio_clear(dev);
+    } else {
+        gpio_set(dev);
+    }
```
Comment (thomaseichinger) [inline cpu/lpc11u34/periph/gpio.c]: I know we are getting into optimisation already, but couldn't this be done by
```C
void gpio_toggle(gpio_t dev)
{
    if (dev >= GPIO_NUMOF) {
        return;
    }
    LPC_GPIO->NOT[gpio_port_map[dev]] = (1 << gpio_pin_map[dev]);
}
```
Saves us one if-statement and two function calls.

**Comment by thomaseichinger** [comment]
@ks156 Great, two minor style comments and one proposal I leave up to your decision. You can squash then right away. It would be great if you could rewrite it to three commits, (1) adding cpu files (2) board files (3) modifying Makefiles.
(You can find information how to use `git rebase` [here](https://git-scm.com/book/en/v2/Git-Tools-Rewriting-History))

**Comment by ks156** [comment]
@thomaseichinger : Changes done. I've also optimized gpio_toggle(). It would have been stupid to leave it as is when the optimization is known and easy to apply.

I'll now try to squash. I think some commits have changed some files in both *cpu* and *board* directories, so it will maybe be difficult to create completely separated commits. I'll see. I have to read and understand how to squash them. As already said, I'm a totally noob with *git rebase*. If needed, is it possible to find someone to help me on IRC ?

**Comment by thomaseichinger** [comment]
You could do it without rebase: (Please backup the RIOT directory in advance to be sure)
```bash
git reset be8f637 # The commit before your first
git add cpu/lpc11u34
git add boards/weio
git commit cpu/lpc11u34/*
git commit boards/weio/*
git commit -a #This should add the Makefile changes
```
I'm in IRC for another hour or so if you have questions.

**Comment by ks156** [comment]
Ok, squashed

**Comment by thomaseichinger** [comment]
Code looks good, Travis succeeded. ACK & Go

**Comment by thomaseichinger** [comment]
@ks156 Congratulations, excellent work. I think this was the fastest PR adding a new board and CPU. From PR open to merge in less than 40 hours. Looking forward to further enhancements.

**Comment by ks156** [comment]
Thank you @thomaseichinger. It was a nice experience for me (first time I create PR for a big project like RIOT). I learned many things these last 40 hours, thanks to your help, advices, and very very good review.

I'll be back soon with more peripherals for this CPU !

**Pull request merged**